mypyc/irbuild/for_helpers.py

```diff
@@ -1180,6 +1190,76 @@ def gen_cleanup(self) -> None:
             gen.gen_cleanup()
 
 
+class ForFilter(ForGenerator):
+    """Generate optimized IR for a for loop over filter(f, iterable)."""
+
+    def need_cleanup(self) -> bool:
+        # The wrapped for loops might need cleanup. We might generate a
+        # redundant cleanup block, but that's okay.
+        return True
+
+    def init(self, index: Lvalue, func: Expression, iterable: Expression) -> None:
+        self.filter_func_def = func
+        if (
+            isinstance(func, NameExpr)
+            and isinstance(func.node, Var)
+            and func.node.fullname == "builtins.None"
+        ):
+            self.filter_func_val = None
+        else:
+            self.filter_func_val = self.builder.accept(func)
+        self.iterable = iterable
+        self.index = index
+
+        self.gen = make_for_loop_generator(
+            self.builder,
+            self.index,
+            self.iterable,
+            self.body_block,
+            self.loop_exit,
+            self.line,
+            is_async=False,
+            nested=True,
+        )
+
+    def gen_condition(self) -> None:
+        self.gen.gen_condition()
+
+    def begin_body(self) -> None:
+        # 1. Assign the next item to the loop variable
+        self.gen.begin_body()
+
+        # 2. Call the filter function
+        builder = self.builder
+        line = self.line
+        item = builder.read(builder.get_assignment_target(self.index), line)
+
+        if self.filter_func_val is None:
+            result = item
+        else:
+            fake_call_expr = CallExpr(self.filter_func_def, [self.index], [ARG_POS], [None])
+
+            # I put this here to prevent a circular import
+            # from mypyc.irbuild.expression import transform_call_expr
+
+            # result = transform_call_expr(builder, fake_call_expr)
+            result = builder.accept(fake_call_expr)
```

Comment:
I'm reverting this to its original state so the tests can turn green again. You'll see this as part of your review anyway.